src/thread_task.rs

```diff
@@ -310,32 +330,33 @@ unsafe fn insert_task(f: *mut Rav1dFrameContext, t: *mut Rav1dTask, cond_signal:
 
 #[inline]
 unsafe fn add_pending(f: *mut Rav1dFrameContext, t: *mut Rav1dTask) {
-    pthread_mutex_lock(&mut (*f).task_thread.pending_tasks.lock);
+    let mut pending_tasks = (*f).task_thread.pending_tasks.lock().unwrap();
     (*t).next = 0 as *mut Rav1dTask;
-    if ((*f).task_thread.pending_tasks.head).is_null() {
-        (*f).task_thread.pending_tasks.head = t;
+    if pending_tasks.head.is_null() {
+        pending_tasks.head = t;
     } else {
-        (*(*f).task_thread.pending_tasks.tail).next = t;
+        (*pending_tasks.tail).next = t;
     }
-    (*f).task_thread.pending_tasks.tail = t;
+    pending_tasks.tail = t;
     (*f).task_thread
         .pending_tasks_merge
         .store(1, Ordering::SeqCst);
-    pthread_mutex_unlock(&mut (*f).task_thread.pending_tasks.lock);
 }
 
 #[inline]
 unsafe fn merge_pending_frame(f: *mut Rav1dFrameContext) -> c_int {
     let merge = (*f).task_thread.pending_tasks_merge.load(Ordering::SeqCst);
     if merge != 0 {
-        pthread_mutex_lock(&mut (*f).task_thread.pending_tasks.lock);
-        let mut t: *mut Rav1dTask = (*f).task_thread.pending_tasks.head;
-        (*f).task_thread.pending_tasks.head = 0 as *mut Rav1dTask;
-        (*f).task_thread.pending_tasks.tail = 0 as *mut Rav1dTask;
-        (*f).task_thread
-            .pending_tasks_merge
-            .store(0, Ordering::SeqCst);
-        pthread_mutex_unlock(&mut (*f).task_thread.pending_tasks.lock);
```

Comment:
`pending_tasks` was previously `unlock`ed/`drop`ped earlier, immediately after `t` is fetched from `pending_tasks`.  We should match that behavior, right?  This does access `t` through a `*mut` after the lock is dropped, though, which I'm not sure will work once `pending_tasks` is made safe, though I think it will since it's removed from the linked list (?) and thus isn't accessible through the shared `pending_tasks` anymore.

Reply:
Should be fixed now.